Use of functions and automatic

I don't think we currently have any guidance on the use of functions, in particular I think this is important for synthesizable  RTL. DV code may want different guidelines or be happy to have none. I'm purely considering synthesizable RTL here.

A few questions to answer:

1. Do we allow use of functions at all in synthesisable RTL?
2. Do we allow use of local variables within functions?
3. Do we mandate use of 'automatic' in all function definitions

I think our answers should be, yes we allow functions, that can use local variables and must always be declared automatic. We probably need some details around how local variables can be used within functions

Clearly functions are useful so saying no functions at all seems very limiting. Allowing local variables though can open up problematic behaviour, e.g.

```systemverilog
function [3:0] do_a_thing(input [3:0] foo);
  logic [3:0] state = 0;
  
  if (foo == ConstantValue) begin
    state = state + 1;
    return state;
  end else begin
    return foo;
  end
endfunction
```

By default `state` is static so this function has persistent state between calls. This has clear issues and shouldn't be allowed.

You can use the 'automatic' qualifier on a function to remove the default static state-holding behaviour I suspect there's some further gotchas around local variables in functions we need to be wary of.

Sorry for being late to the party here. 3x times yes as well from my side. I just wanted to add that lint would flag non-automatic functions in RTL. For example, if I remove `automatic` from one of the functions in `aes_pkg.sv`, the following error is thrown:
```
E   AUTOMATIC:   aes_pkg.sv:69   Function 'aes_mul2' is not declared as 'automatic'
```

--- a/VerilogCodingStyle.md
+++ b/VerilogCodingStyle.md
@@ -103,6 +103,7 @@ representation of this style guide.
     - [Generate Constructs](#generate-constructs)
     - [Signed Arithmetic](#signed-arithmetic)
     - [Number Formatting](#number-formatting)
+    - [Functions and Tasks](#functions-and-tasks)
     - [Problematic Language Features and Constructs](#problematic-language-features-and-constructs)
       - [Floating begin-end blocks](#floating-begin-end-blocks)
   - [Design Conventions](#design-conventions)
@@ -2289,6 +2290,156 @@ always_comb begin
     addr2 = 40'h41_c000_0000;
   end
 end
+```
+
+### Functions and Tasks
+
+The following section applies to synthesizable RTL only. See the [Coding Style
+Guide for Design Verification](DVCodingStyle.md) for DV usage.
+
+***In synthesizable RTL the use of functions is allowed, provided they are declared
+`automatic`. Tasks should not be used.***
+
+Functions must be declared in either a package or inside a module. A package is
+appropriate where the function relates to other definitions in the package and
+could be useful to multiple modules (even if it's currently only used by one). A
+module is appropriate where the function specifically relates to the internals
+of that module.
+
+Functions should aim to conceptually represent a reusable block of combinational
+logic.
+
+Storage types must be explicitly declared for all arguments and the function
+return value. All types must be 4-state data types, either `logic` or types
+derived from `logic` (such as appropriate `struct`, `enum` or `typedef` types).
+
+Do not use `output`, `inout`, or `ref` on function arguments. All functions
+should only consume inputs and produce one output. `input` is the default and is
+not required on the function arguments.
+
+
+&#x1f44e;
+```systemverilog {.bad}
+// - Doesn't have explicit storage type on `a` or `b` or return type
+// - `b` being used as `output` argument
+// - `input` not required on `a`
+function automatic [2:0] foo(input [2:0] a, [2:0] b);
+  b = b + 1;
+  return a + b;
+endfunction
+```
+
+&#x1f44e;
+```systemverilog {.bad}
+// - Doesn't have explicit storage type on `a`, `b` or `c`
+// - Uses `output` on `c`
+// - `input` not required on `a` and `b`
+function automatic logic [2:0] foo(input [2:0] a, input [2:0] b, output [2:0] c);
+  c = a - b;
+  return a + b;
+endfunction
+```
+
+&#x1f44e;
+```systemverilog {.bad}
+// - Uses 2-state data type `int` for `a`
+function automatic logic [2:0] foo(int a, logic [2:0] b);
+  return a + b;
+endfunction
+```
+
+&#x1f44d;
+```systemverilog {.good}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+ return a ^ b;
+endfunction
+```
+
+&#x1f44d;
+```systemverilog {.good}
+typedef logic [2:0] bar_t;
+
+typedef struct packed {
+  logic [2:0] field;
+} baz_t;
+
+function automatic logic [2:0] foo(bar_t a, baz_t b);
+  return a + b.field;
+endfunction
+```
+
+Data should be returned from a function using an explicit `return result` style.
+Do not use a `function_name = result` style.
+
+&#x1f44e;
+```systemverilog {.bad}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+  if (a == 3'd2) begin
+    foo = b;
+  end else begin
+    foo = a ^ b;
+  end
+endfunction
+```
+
+&#x1f44d;
+```systemverilog {.good}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+  logic [2:0] result;
+
+  if (a == 3'd2) begin
+    result = b;
+  end else begin
+    result = a ^ b;
+  end
+
+  return result;
+endfunction
+```
+
+All local variables must be assigned in all code paths, either through an
+initial assignment or through the use of `else` and `default:` for `if` and
+`case` statements.
+
+&#x1f44d;
+```systemverilog {.good}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+  logic [2:0] local_var_1;
+  logic [2:0] local_var_2;
+
+  local_var_1 = 3'd0;
+
+  if (a == 0) begin
+    local_var_1 = 3'd2;
+  end
+
+  unique case(b)
+    3'd0:    local_var_2 = 3'd1;
+    3'd1:    local_var_2 = 3'd3;
+    default: local_var_2 = 3'd0;
+  endcase
+
+  return local_var_1 + local_var_2;
+endfunction
+```
+
+&#x1f44e;
+```systemverilog {.bad}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+  logic [2:0] local_var_1;
+  logic [2:0] local_var_2;
+
+  if (a == 0) begin
+    local_var_1 = 3'd2;
+  end
+
+  unique case(b)
+    3'd0:    local_var_2 = 3'd1;
+    3'd1:    local_var_2 = 3'd3;
+  endcase
+
+  return local_var_1 + local_var_2;
+endfunction
 ```
 
 ### Problematic Language Features and Constructs

